tests/test_make_caster_adl.cpp

```diff
@@ -0,0 +1,92 @@
+// adl = Argument Dependent Lookup
+
+#include "pybind11_tests.h"
+
+namespace have_a_ns {
+struct mock_caster {
+    static int num() { return 101; }
+};
+struct type_mock {
+    friend mock_caster pybind11_select_caster(type_mock *);
+};
+
+} // namespace have_a_ns
+
+// namespace global {
+struct global_ns_type_mock {};
+struct global_ns_mock_caster {
+    static int num() { return 202; }
+};
+global_ns_mock_caster pybind11_select_caster(global_ns_type_mock *);
+// } // namespace global
+
+namespace {
+struct unnamed_ns_type_mock {};
+struct unnamed_ns_mock_caster {
+    static int num() { return 303; }
+};
+unnamed_ns_mock_caster pybind11_select_caster(unnamed_ns_type_mock *);
+#if !defined(_MSC_VER)
+// Dummy implementation, purely to avoid compiler warnings (unused function).
+// Easier than managing compiler-specific pragmas for warning suppression.
+// MSVC happens to not generate any warnings. Leveraging that to prove that
+// this test actually works without an implementation.
```

Comment:
Maybe mention `[[maybe_unused]]` for C++17? If we ever go C++17+ only, we can then just use that.

Reply:
Thanks for the suggestion! We already have `PYBIND11_MAYBE_UNUSED`, trying that now in the CI.